addons/website_slides/controllers/main.py

```diff
@@ -554,6 +561,7 @@ def slide_set_completed(self, slide_id):
         if request.website.is_public_user():
             return {'error': 'public_user'}
         fetch_res = self._fetch_slide(slide_id)
+        self._slide_set_completed_check(fetch_res['slide'])
```

Comment:
Few things about this method:

- If you leave it like that, please put the check after "if fetch_res.get('error'):"

- If I may suggest, wouldn't it be easier to implement it like that:

```
def _set_completed_slide(self, slide, quiz_attempts_inc=False, quiz_done=False):
        if slide.website_published and slide.channel_id.is_member:
            if (slide.slide_type == 'quiz' or slide.question_ids) and not quiz_done:
            raise werkzeug.exceptions.Forbidden(_("Slide with questions must be marked as done when submitting all good answers "))
            slide.action_set_completed(quiz_attempts_inc=quiz_attempts_inc)
        return True
```

And in "slide_quiz_submit":

```
...
self._set_completed_slide(slide, quiz_attempts_inc=True, quiz_done=True)
...
```

Tell me what you think.

Reply:
But I need to prevent the set_completed route for certification. So, I need to extend the check.
The idea of doing this in the helper in the controller is cool, but I 'll need to call the method object `action_set_completed` directly in the /submit route. All this code is a little spaghetti ...